### PR TITLE
Use mock from standard library

### PR DESCRIPTION
**Pull request by V02460**
~~As Python 3.6 is EOL, the backported version of the mock package is no longer needed and can be replaced by the one found in the standard library. Workarounds for emulating `AsyncMock` are no longer required either.~~

Synapse currently uses the backported non-standard library mock in one single test. This PR rewrites that test to no longer require the non-standard mock and removes the dependency from Synapse.

### CODE REVIEW

**Comment by clokep** [comment]
CI will fail on this until we remove Python 3.6, unfortunately.

I suspect we can also replace `make_awaitable`, but I'm not 100% sure on that one!

**Comment by DMRobertson** [comment]
Related: #11588

**Comment by clokep** [comment]
This looks good overall -- I'm removing this from the review queue for now since it is blocked on our CI dropping support for Python 3.6.

**Comment by callahad** [comment]
(Note: that we _are_ open to PRs that drop 3.6 from our CI pipeline now that 1.49.0 is out)

**Comment by H-Shay** [comment]
There is a PR open now for dropping 3.6 from CI: #11595

**Comment by V02460** [comment]
Some corrections: I just noticed that `AsyncMock` is not available in Python 3.7 either, so removing `simple_async_mock` has to wait for another 1½ years >.<

Looking at the code now, it seems you went through this discussion before, but there is currently a single test in Synapse which requires a backported `Mock` on Python 3.7 (tests.storage.test_background_update.BackgroundUpdateControllerTestCase.test_controller). If that instance is changed, mock could still be removed – unless you want to keep your options.

**Comment by V02460** [comment]
CI completes successfully now, even with Python 3.6!

File: tests/storage/test_background_update.py [hunk 1]
```diff
@@ -116,14 +115,15 @@ def prepare(self, reactor, clock, homeserver):
...
-        self._update_ctx_manager.__aenter__ = Mock(
-            return_value=make_awaitable(None),
-        )
-        self._update_ctx_manager.__aexit__ = Mock(return_value=make_awaitable(None))
+        class MockCM:
+            pass
+
+        self._update_ctx_manager = MockCM
+        self._update_ctx_manager.__aenter__ = simple_async_mock(return_value=None)
+        self._update_ctx_manager.__aexit__ = simple_async_mock(return_value=None)
```
Comment (clokep) [inline tests/storage/test_background_update.py:123]: Would it work to define these on the class directly? That seems a bit cleaner!

Comment (V02460) [reply tests/storage/test_background_update.py]: Yes!

**Review by clokep — APPROVED** [review]
That test seems to be rather confusing, but this doesn't make it anymore so. Thanks for taking care of this! 👍

**Pull request merged**